### DISCUSSION TITLE
Is `setuptools.find_namespace_packages()` consistent with PEP420?

### DISCUSSION THREAD

**Discussion by kandersolar** [pypa/setuptools#3427] [Q&A] [answered]
I'm trying to understand the new best practice for package maintainers around PEP 420 and found something that seems inconsistent between how `setuptools` defines namespace packages and how PEP 420 defines them.

Under https://peps.python.org/pep-0420/#specification it says:

> Namespace packages cannot contain an \_\_init__.py.

And under [Differences between namespace packages and regular packages](https://peps.python.org/pep-0420/#differences-between-namespace-packages-and-regular-packages), PEP 420 says:

> Namespace packages have no \_\_init__.py module.

From this I'd have thought that a namespace package MUST NOT (in the convention of [RFC 2119](https://datatracker.ietf.org/doc/html/rfc2119)) contain an `__init__.py` module.  However, the code underlying `find_namespace_packages` does not reflect that requirement and considers any directory a namespace package, regardless of the presence or absence of an `__init__.py` module:

https://github.com/pypa/setuptools/blob/641cb8955923f5b23c5be7cdd3edb7427c40995c/setuptools/discovery.py#L163-L166

I've found very helpful explanations from @abravalheri (https://github.com/pypa/setuptools/issues/3340#issuecomment-1146676266 and elsewhere) about the definition of packages, but I've not seen this particular aspect of namespace packages discussed.  Any clarification here would be appreciated :)

**Comment by abravalheri** [comment, marked as answer]
Hi @kanderso-nrel ,maybe the naming we use right now is a bit mixed up, but the idea with `find_namespace_packages` is that it would *also include namespaces* in addition to regular packages. The same thing with `PEP420PackageFinder`, it find the packages including the ones described in PEP 420.

**Reply by kandersolar** [reply]
Thanks for the explanation @abravalheri!  From their names alone I might have assumed that `find_packages` and `find_namespace_packages` would have no overlap in what they return, i.e. `find_packages` returns _only_ directories with an `__init__` file and `find_namespace_packages` returns _only_ directories without one.  Thanks for clarifying.

**Reply by abravalheri** [reply]
Yeah, the choice of names might have been a bit misleading, but since these functions/classes are part of setuptools for a long time now, I don't think it would be wise to rename them. What we can do is improve documentation. Please let us know if they appear in a part of the docs that is confusing.
